lib/charms/mongodb/v0/mongo.py

```diff
@@ -0,0 +1,238 @@
+"""Code for interactions with MongoDB."""
+
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+import abc
+import logging
+import re
+from dataclasses import dataclass
+from typing import List, Optional, Set
+
+from pymongo import MongoClient
+from pymongo.errors import OperationFailure, PyMongoError
+from tenacity import RetryError, Retrying, stop_after_delay, wait_fixed
+
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+
+class NotReadyError(PyMongoError):
+    """Raised when not all replica set members healthy or finished initial sync."""
+
+
+# The unique Charmhub library identifier, never change it
+LIBID = "3037662a76cc4bf1876d4659c88e77e5"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+# path to store mongodb ketFile
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class MongoConfiguration:
+    """Class for Mongo configurations usable my mongos and mongodb.
+
+    — replset: name of replica set
+    - port: connection port
+    — database: database name.
+    — username: username.
+    — password: password.
+    — hosts: full list of hosts to connect to, needed for the URI.
+    - tls_external: indicator for use of internal TLS connection.
+    - tls_internal: indicator for use of external TLS connection.
+    """
+
+    database: Optional[str]
+    username: str
+    password: str
+    hosts: Set[str]
+    roles: Set[str]
+    tls_external: bool
+    tls_internal: bool
+
+    @property
+    @abc.abstractmethod
+    def uri(self):
+        """Return URI concatenated from fields."""
```

Comment:
Even though it's not mandatory, you can add a `raise NotImplentedError` here to ensure that it's implemented in the child classes.
This way you enforce the subclasses to implement this method.

Reply:
🧠